File: lib/models/market_data.dart
class FuturesPrice {
  final String id;
  final String symbol;
  final String name;
  final double price;
  final double change;
  final double changePercent;
  final DateTime lastUpdated;

  FuturesPrice({
    required this.id,
    required this.symbol,
    required this.name,
    required this.price,
    required this.change,
    required this.changePercent,
    required this.lastUpdated,
  });

  String get formattedPrice => '\$${price.toStringAsFixed(2)}';

  String get formattedChange {
    final sign = change >= 0 ? '+' : '';
    return '$sign${change.toStringAsFixed(2)}';
  }

  String get formattedChangePercent {
    final sign = changePercent >= 0 ? '+' : '';
    return '$sign${changePercent.toStringAsFixed(2)}%';
  }

  bool get isPositive => change >= 0;
}

class PriceDataPoint {
  final String id;
  final DateTime date;
  final double price;

  PriceDataPoint({
    required this.id,
    required this.date,
    required this.price,
  });
}

class AuctionReport {
  final String id;
  final DateTime reportDate;
  final String marketLocation;
  final String state;
  final int headCount;
  final List<WeightClassPrice> weightClasses;

  AuctionReport({
    required this.id,
    required this.reportDate,
    required this.marketLocation,
    required this.state,
    required this.headCount,
    required this.weightClasses,
  });
}

class WeightClassPrice {
  final String id;
  final String weightRange;
  final double averagePrice;
  final double lowPrice;
  final double highPrice;
  final int headCount;

  WeightClassPrice({
    required this.id,
    required this.weightRange,
    required this.averagePrice,
    required this.lowPrice,
    required this.highPrice,
    required this.headCount,
  });

  String get formattedAverage => '\$${averagePrice.toStringAsFixed(2)}';

  String get formattedRange =>
      '\$${lowPrice.toStringAsFixed(2)} - \$${highPrice.toStringAsFixed(2)}';
}

class MarketSummary {
  final FuturesPrice? feederCattle;
  final FuturesPrice? liveCattle;
  final List<AuctionReport> recentAuctions;
  final List<PriceDataPoint> priceHistory;
  final DateTime lastUpdated;

  MarketSummary({
    this.feederCattle,
    this.liveCattle,
    this.recentAuctions = const [],
    this.priceHistory = const [],
    required this.lastUpdated,
  });

  static MarketSummary get empty => MarketSummary(
        lastUpdated: DateTime.now(),
      );
}

File: swift_reference/Models/MarketData.swift
import Foundation

struct FuturesPrice: Identifiable, Codable {
    let id: UUID
    let symbol: String
    let name: String
    let price: Double
    let change: Double
    let changePercent: Double
    let lastUpdated: Date

    init(
        id: UUID = UUID(),
        symbol: String,
        name: String,
        price: Double,
        change: Double,
        changePercent: Double,
        lastUpdated: Date = Date()
    ) {
        self.id = id
        self.symbol = symbol
        self.name = name
        self.price = price
        self.change = change
        self.changePercent = changePercent
        self.lastUpdated = lastUpdated
    }

    var formattedPrice: String {
        String(format: "$%.2f", price)
    }

    var formattedChange: String {
        let sign = change >= 0 ? "+" : ""
        return "\(sign)\(String(format: "%.2f", change))"
    }

    var formattedChangePercent: String {
        let sign = changePercent >= 0 ? "+" : ""
        return "\(sign)\(String(format: "%.2f", changePercent))%"
    }

    var isPositive: Bool {
        change >= 0
    }
}

struct PriceDataPoint: Identifiable {
    let id = UUID()
    let date: Date
    let price: Double
}

struct AuctionReport: Identifiable, Codable {
    let id: UUID
    let reportDate: Date
    let marketLocation: String
    let state: String
    let headCount: Int
    let weightClasses: [WeightClassPrice]

    init(
        id: UUID = UUID(),
        reportDate: Date,
        marketLocation: String,
        state: String,
        headCount: Int,
        weightClasses: [WeightClassPrice]
    ) {
        self.id = id
        self.reportDate = reportDate
        self.marketLocation = marketLocation
        self.state = state
        self.headCount = headCount
        self.weightClasses = weightClasses
    }
}

struct WeightClassPrice: Identifiable, Codable {
    let id: UUID
    let weightRange: String
    let averagePrice: Double
    let lowPrice: Double
    let highPrice: Double
    let headCount: Int

    init(
        id: UUID = UUID(),
        weightRange: String,
        averagePrice: Double,
        lowPrice: Double,
        highPrice: Double,
        headCount: Int
    ) {
        self.id = id
        self.weightRange = weightRange
        self.averagePrice = averagePrice
        self.lowPrice = lowPrice
        self.highPrice = highPrice
        self.headCount = headCount
    }

    var formattedAverage: String {
        String(format: "$%.2f", averagePrice)
    }

    var formattedRange: String {
        "$\(String(format: "%.2f", lowPrice)) - $\(String(format: "%.2f", highPrice))"
    }
}

struct MarketSummary {
    let feederCattle: FuturesPrice?
    let liveCattle: FuturesPrice?
    let recentAuctions: [AuctionReport]
    let priceHistory: [PriceDataPoint]
    let lastUpdated: Date

    static let empty = MarketSummary(
        feederCattle: nil,
        liveCattle: nil,
        recentAuctions: [],
        priceHistory: [],
        lastUpdated: Date()
    )
}

struct CattleDrug: Identifiable {
    let id = UUID()
    let name: String
    let dosagePerCwt: Double
    let unit: String
    let route: TreatmentRoute
    let withdrawalDays: Int
    let notes: String

    var dosageDescription: String {
        "\(dosagePerCwt) \(unit) per 100 lbs"
    }

    static let commonDrugs: [CattleDrug] = [
        CattleDrug(name: "Draxxin", dosagePerCwt: 1.1, unit: "mL", route: .subcutaneous, withdrawalDays: 28, notes: "Tulathromycin - BRD treatment"),
        CattleDrug(name: "Excede", dosagePerCwt: 1.0, unit: "mL", route: .subcutaneous, withdrawalDays: 13, notes: "Ceftiofur - BRD treatment"),
        CattleDrug(name: "Nuflor", dosagePerCwt: 3.0, unit: "mL", route: .subcutaneous, withdrawalDays: 28, notes: "Florfenicol - BRD treatment"),
        CattleDrug(name: "Baytril 100", dosagePerCwt: 3.4, unit: "mL", route: .subcutaneous, withdrawalDays: 28, notes: "Enrofloxacin - BRD treatment"),
        CattleDrug(name: "Micotil", dosagePerCwt: 1.5, unit: "mL", route: .subcutaneous, withdrawalDays: 28, notes: "Tilmicosin - BRD treatment"),
        CattleDrug(name: "LA-200", dosagePerCwt: 4.5, unit: "mL", route: .intramuscular, withdrawalDays: 28, notes: "Oxytetracycline - Various infections"),
        CattleDrug(name: "Dectomax", dosagePerCwt: 1.0, unit: "mL", route: .subcutaneous, withdrawalDays: 35, notes: "Doramectin - Parasiticide"),
        CattleDrug(name: "Ivomec", dosagePerCwt: 1.0, unit: "mL", route: .subcutaneous, withdrawalDays: 35, notes: "Ivermectin - Parasiticide"),
        CattleDrug(name: "Safe-Guard", dosagePerCwt: 2.3, unit: "mL", route: .oral, withdrawalDays: 8, notes: "Fenbendazole - Dewormer"),
        CattleDrug(name: "Banamine", dosagePerCwt: 1.0, unit: "mL", route: .intravenous, withdrawalDays: 4, notes: "Flunixin - Anti-inflammatory"),
        CattleDrug(name: "Dexamethasone", dosagePerCwt: 1.0, unit: "mL", route: .intramuscular, withdrawalDays: 0, notes: "Corticosteroid - Anti-inflammatory"),
        CattleDrug(name: "Penicillin G", dosagePerCwt: 3.0, unit: "mL", route: .intramuscular, withdrawalDays: 10, notes: "Antibiotic - Various infections")
    ]
}
